### ISSUE TITLE
Place for JS Logic for Route-less Engines

### ISSUE THREAD

**Issue by sangm** [ember-engines/ember-engines#389]
As I'm playing around with Route less engines, I'm not exactly sure where the JS logic should go.

@mike183 @trentmwillis @rwjblue Can you guys chime in?

**Comment by knownasilya** [comment]
The logic would go into components, which would be setup in the application template.

**Comment by trentmwillis** [comment]
That is definitely how one would need to do it currently, but there are still a few unsolved questions, such as how would you access params passed into `{{mount}}` (a feature we will likely add)? Also, having an intermediary template just to have a component feels a bit awkward IMO.

My suggestion, and I've discussed this with @mike183 briefly, is that there needs to be a backing JS class that is associated with a route-less engine's `application.hbs` (maybe `application.js`?). The actual construct should probably be similar to what is used by components. This provides intuitive and already well-defined hooks for working with params (e.g., `didReceiveAttrs`), but would likely want to have first-class support for asynchronously computing values (e.g., fetching data).

I think someone more familiar with how Components and Routes currently work could speak to implementation details better, but this is what I think needs to be done from a high-level.

**Comment by sangm** [comment]
Agreed. Based on intuition, if `application.hbs` is what gets rendered, I expect `application.js` to execute the logic.

**Comment by mike183** [comment]
> Also, having an intermediary template just to have a component feels a bit awkward IMO.

Agreed, just seems like boilerplate code that a vast majority of people creating routeless engines will need to do.

**Comment by knownasilya** [comment]
This might be the place to experiment with the routable-component "route" interface.
Since the `application.js` file is basically a route since it has params and should be able to provide data.

**Comment by trentmwillis** [comment]
> This might be the place to experiment with the routable-component "route" interface.

This is basically what I intend, but I want to steer clear of association with routable components haha.
